lib/vdc/Grid.cpp

```diff
@@ -141,24 +142,55 @@ void Grid::SetValueIJK(size_t i, size_t j, size_t k, float v)
 
 void Grid::GetRange(float range[2]) const
 {
-    float               missingValue = GetMissingValue();
-    Grid::ConstIterator itr = this->cbegin();
-    Grid::ConstIterator enditr = this->cend();
+    const auto missingVal = GetMissingValue();
+    const auto num_vals = _dims[0] * _dims[1] * _dims[2];
+    auto num_threads = size_t{0};
+#pragma omp parallel
+    {
+      if (omp_get_thread_num() == 0)
+        num_threads = omp_get_num_threads();
+    }
+    assert(num_vals > num_threads);
+    const auto stride_size = num_vals / num_threads;
+
+    auto min_vec = std::vector<float>(num_threads, missingVal);
+    auto max_vec = std::vector<float>(num_threads, missingVal);
+
+#pragma omp parallel for
+    for (size_t i = 0; i < num_threads; i++) {
+      auto beg = this->cbegin() + i * stride_size;
+      auto end = beg;
+      if (i < num_threads - 1) 
+          end += stride_size;
+      else
+          end = this->cend();
+
+      while (*beg == missingVal && beg != end) { ++beg; }
+
+      if (beg != end) {
+        auto min = *beg;
+        auto max = *beg;
+        for (auto itr  = beg + 1; itr != end; ++itr) {
+          if (*itr != missingVal) {
+            min = std::min(*itr, min);      
```

Comment:
On my Mac replacing the use of the STL operators with:

```
if (*itr < min) min = *itr;
if (*itr > max) max = *itr;
```

Runs 10 to 15% faster.

Reply:
Interesting. Maybe it's a platform/data type specific thing, but STL algorithms such as 'std::min()' are optimized to use certain hardware instructions instead of if/else conditions, so they're supposed to be faster. I think I'll keep this one here ;)

Reply:
@shaomeng I think you are comparing apples to oranges :-). If the code were:

`min = a < b ? a : b`

I would agree that std::min() should be just as efficient (that would be an apple to apple comparison). But the code in question does something different: it only *conditionally* assigns a new minimum value. Thus, there are fewer assignments than with your implementation. I ran this on my may with clang12. Try it on another platform/compiler and see if you don't get similar results.

Reply:
Interesting. I tried on cisl-vapor (Ubuntu 20, GCC 9.3) and std::min() is consistently faster. In the following screenshot, the first 3 runs with time ~2.1s are using if/else, and the following 3 runs with time ~1.8s are using std::min().
<img width="500" alt="Screen Shot 2022-04-26 at 11 23 21 AM" src="https://user-images.githubusercontent.com/814885/165357802-54fbc43c-97ec-4122-98b9-9b17e26e58a8.png">

On the other hand, I tested a super simple standalone program, and there's virtually no difference... I guess it has to be tested on the specific program one's using...

Reply:
@shaomeng, just to be clear, you said you are comparing against "if/else". The code snippet I shared wasn't an "if/else", it was an "if". There is a difference because there is no "else" execution branch. So are you comparing std::min against the "if" that I shared, or the "if/else"?  Thanks.

Reply:
Sorry, I was using "if/else" as a name for the conditional pattern, but the actual test code didn't involve an 'else' part in this case. It was a comparison between `min = std::min(val, min)` and `if (val < min) min = val;`. 

I think it ultimately comes down to if it's faster to do an assignment or to resolve a condition. It could well be a platform dependent thing...

Reply:
Got it. Thanks.